configure

```diff
@@ -292,277 +332,749 @@ LIBS="${LIBS} mpfr"
+            BUILD_MACHINE="aarch64"
+            ;;
+        arm*)
+            BUILD_MACHINE="arm"
+            ;;
+        ia64)
+            BUILD_MACHINE="ia64"
+            ;;
+        sparc | sun4*)
+            BUILD_MACHINE="sparc"
+            ;;
+        sparc64)
+            BUILD_MACHINE="sparc64"
+            ;;
+        ppc64 | powerpc64)
+            BUILD_MACHINE="ppc64"
+            ;;
+        ppc | powerpc | [P|p]ower*)
+            BUILD_MACHINE="ppc"
+            ;;
+        mips64)
+            BUILD_MACHINE="mips64"
+            ;;
+        *)
+            BUILD_MACHINE="unknown"
+            ;;
+    esac
+
+    ############################################################################
+    # OS / operating system
+    ############################################################################
+    if [ "$(uname | cut -d_ -f1)" = "MSYS" ]; then
+        if [ "$BUILD_ARCH" = "x86_64" ]; then
+            BUILD_OS="MINGW64"
+        else
+            BUILD_OS="MINGW32"
+        fi
+    elif [ "$(uname | cut -d_ -f1)" = "MINGW32" ]; then
+        if [ "$ABI" = "64" ]; then
+            BUILD_OS="MINGW64"
+        else
+            BUILD_OS="MINGW32"
+        fi 
+    elif [ "$(uname | cut -d_ -f1)" = "MINGW64" ]; then
+        if [ "$ABI" = "32" ]; then
+            BUILD_OS="MINGW32"
+        else
+            BUILD_OS="MINGW64"
+        fi
+    elif [ "$(uname | cut -d_ -f1)" = "CYGWIN" ]; then
+        if [ "$BUILD_ARCH" = "x86_64" ]; then
+            if [ "$ABI" = "32" ]; then
+                BUILD_OS="CYGWIN32"
+            else
+                BUILD_OS="CYGWIN64"
+                ABI="64"
+            fi
+        else
+            BUILD_OS="CYGWIN32"
+        fi
+    else
+        BUILD_OS=`uname -s`
+    fi
 else
-   ARCH=`echo "$BUILD" | cut -d- -f1`
-   OS=`echo "$BUILD" | cut -d- -f2`
+    # Build was given
+
+    ############################################################################
+    # architecture and machine
+    ############################################################################
+    case $BUILD in
+        ########################################################################
+        # x86
+        ########################################################################
+        i?86*-* | pentium*-* | atom-* | prescott-* | nocona-* | bonnell-* |\
+        silvermont-* | goldmont*-* | tremont-* | core2-* | corei*-* |\
+        nehalem*-* | westmere*-* | sandybridge*-* | ivybridge*-* | haswell*-* |\
+        broadwell*-* | skylake*-* | kabylake*-* | coffeelake*-* | knl*-* |\
+        knm*-* | cascadelake*-* | cannonlake*-* | cooperlake*-* | icelake*-* |\
+        tigerlake*-* | sapphirerapids*-* | rocketlake*-* | alderlake*-* |\
+        k[5-8]*-* | geode*-* | athlon*-* | opteron*-* | k10-* | amdfam10-* |\
+        barcelona-* | bt*-* | bobcat-* | jaguar*-* | bd*-* | bulldozer*-* |\
+        piledriver*-* | steamroller*-* | excavator*-* | zn*-* | zen*-* |\
+        x86_64*-*)
```

Comment:
I don't understand -- why would anyone want to specify such a BUILD string, instead of just directly giving a list of desired `-mFOO` flags? A major downside is that that this list would require constant maintenance to keep it up-to-date. But what are the upsides?

Reply:
It it what GMP does. But if someone is good at Autoconf, I would really appreciate some help.

Reply:
The upside would be architecture specific assembly optimisations. That could be useful for a few things, though I don't know how currently urgent it is.